pkg/chunk/cassandra/scanner.go

```diff
@@ -0,0 +1,146 @@
+package cassandra
+
+import (
+	"context"
+	"fmt"
+	"sync"
+
+	"github.com/cortexproject/cortex/pkg/chunk"
+	"github.com/cortexproject/cortex/pkg/chunk/cassandra"
+	"github.com/pkg/errors"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/common/model"
+	"github.com/sirupsen/logrus"
+)
+
+// scanBatch represents a batch of rows read from Cassandra.
+type scanBatch struct {
+	hash       []byte
+	rangeValue []byte
+	value      []byte
+}
+
+type IndexValidator struct {
+	schema chunk.SchemaConfig
+	s      *StorageClient
+	o      *ObjectClient
+}
+
+func NewIndexValidator(
+	cfg cassandra.Config,
+	schema chunk.SchemaConfig,
+) (*IndexValidator, error) {
+	o, err := NewObjectClient(
+		cfg,
+		schema,
+		prometheus.NewRegistry(),
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	s, err := NewStorageClient(
+		cfg,
+		schema,
+		prometheus.NewRegistry(),
+	)
+	if err != nil {
+		return nil, err
+	}
+	return &IndexValidator{schema, s, o}, nil
+}
+
+func (i *IndexValidator) IndexScan(ctx context.Context, table string, from model.Time, to model.Time, out chan string) error {
+	q := i.s.readSession.Query(fmt.Sprintf("SELECT hash, range, value FROM %s", table))
+
+	iter := q.WithContext(ctx).Iter()
+	defer iter.Close()
+	scanner := iter.Scanner()
+
+	wg := &sync.WaitGroup{}
+	batchChan := make(chan scanBatch)
+
+	for n := 0; n < 16; n++ {
+		wg.Add(1)
+		go func() {
+			defer wg.Done()
+			for b := range batchChan {
+				i.checkEntry(ctx, from, to, out, b)
+			}
+		}()
+	}
+
+	rowsReadTotal := 0
+
+	for scanner.Next() {
+		b := scanBatch{}
+		if err := scanner.Scan(&b.hash, &b.rangeValue, &b.value); err != nil {
+			return errors.WithStack(err)
+		}
+		batchChan <- b
+		rowsReadTotal++
+		if rowsReadTotal%1000 == 0 {
+			logrus.Infof("index entries scanned total: %d\n", rowsReadTotal)
+		}
+	}
+	close(batchChan)
+	wg.Wait()
+	return errors.WithStack(scanner.Err())
+}
+
+func (i *IndexValidator) checkEntry(
+	ctx context.Context,
+	from model.Time,
+	to model.Time,
+	out chan string,
+	entry scanBatch,
+) {
+	chunkID, _, isSeriesID, err := parseChunkTimeRangeValue(entry.rangeValue, entry.value)
+	if err != nil {
+		logrus.WithField("chunk_id", chunkID).WithError(err).Errorln("unable to parse chunk time range value")
+		return
+	}
+
+	if isSeriesID {
+		logrus.WithField("series_id", chunkID).Debugln("ignoring series id row")
+		return
+	}
+
+	c, err := chunk.ParseExternalKey("fake", chunkID)
```

Comment:
should this be hard-coded to `fake`?

Reply:
Good catch, it should be configurable but default to fake